File: FrontEnd_V2/src/api/goals.ts
import { ENDPOINTS } from "@/constant/shadow-endpoints";

import { http } from "@/api/client";
import {
    GoalDataResponse,
    GoalDataShortResponse,
    GoalListStatusFilter, 
    RefineGoalResponse,
    RefineGoalRequest, 
    RefineGoalFromLLMSchema
} from "@/api/types";

export const goalsApi = {
    async understandGoal(data: RefineGoalRequest): Promise<RefineGoalResponse> {
        return http.post<RefineGoalResponse>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.REFINE}`, data);
    },

    async saveGoal(data: RefineGoalFromLLMSchema): Promise<void> {
        return http.post<void>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.SAVE}`, data);
    },

    async getList(status: GoalListStatusFilter): Promise<GoalDataShortResponse[]> {
        return http.get<GoalDataShortResponse[]>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.GET_LIST}`, {
            params: { status },
        });
    },

    async getDetail(goalId: number): Promise<GoalDataResponse> {
        return http.get<GoalDataResponse>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.DETAIL(goalId)}`);
    },

    async deleteGoal(goalId: number): Promise<void> {
        return http.delete<void>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.DETAIL(goalId)}`);
    },

    async updateGoal(goalId: number, data: RefineGoalFromLLMSchema): Promise<GoalDataResponse> {
        return http.patch<GoalDataResponse>(`${ENDPOINTS.GOALS.PREFIX}${ENDPOINTS.GOALS.DETAIL(goalId)}`, data);
    },
};

File: BackEnd_V2/app/api/goals.py
from fastapi import APIRouter, Depends, status as http_status

from app.llm import RefineGoalFromLLM
from app.api.deps import get_current_user
from app.core.endpoints import ENDPOINTS
from app.db.session import get_db
from app.models.user import UserDBM
from app.schemas.goals import (
    GoalDataResponse,
    GoalDataShortResponse,
    GoalListStatusFilter,
    SaveGoalRequest,
    RefineGoalRequest,
)
from app.services import goals_service

router = APIRouter(prefix=ENDPOINTS.GOALS.PREFIX, tags=["Goals"])


@router.post(
    ENDPOINTS.GOALS.REFINE,
    response_model=RefineGoalFromLLM,
)
async def refine_goal(
    data: RefineGoalRequest,
    current_user: UserDBM = Depends(get_current_user),
) -> RefineGoalFromLLM:
    return await goals_service.refine_goal(data, current_user)


@router.post(ENDPOINTS.GOALS.SAVE, status_code=http_status.HTTP_204_NO_CONTENT)
def save_goal(
    data: SaveGoalRequest,
    db=Depends(get_db),
    current_user: UserDBM = Depends(get_current_user),
) -> None:
    goals_service.save_goal(db, current_user, data)


@router.get(ENDPOINTS.GOALS.GET_LIST, response_model=list[GoalDataShortResponse])
def get_goal_list(
    status: GoalListStatusFilter = "All",
    db=Depends(get_db),
    current_user: UserDBM = Depends(get_current_user),
) -> list[GoalDataShortResponse]:
    return goals_service.get_goal_list(db, current_user, status)


@router.get(ENDPOINTS.GOALS.DETAIL, response_model=GoalDataResponse)
def get_goal_detail(
    goal_id: int,
    db=Depends(get_db),
    current_user: UserDBM = Depends(get_current_user),
) -> GoalDataResponse:
    return goals_service.get_goal_detail(db, current_user, goal_id)


@router.delete(ENDPOINTS.GOALS.DETAIL, status_code=http_status.HTTP_204_NO_CONTENT)
def delete_goal(
    goal_id: int,
    db=Depends(get_db),
    current_user: UserDBM = Depends(get_current_user),
) -> None:
    goals_service.delete_goal(db, current_user, goal_id)


@router.patch(ENDPOINTS.GOALS.DETAIL, response_model=GoalDataResponse)
def update_goal(
    goal_id: int,
    data: SaveGoalRequest,
    db=Depends(get_db),
    current_user: UserDBM = Depends(get_current_user),
) -> GoalDataResponse:
    return goals_service.update_goal(db, current_user, goal_id, data)
